game-core/src/main/java/games/strategy/triplea/delegate/power/calculator/AaDefenseCombatValue.java

```diff
@@ -0,0 +1,95 @@
+package games.strategy.triplea.delegate.power.calculator;
+
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.Unit;
+import games.strategy.triplea.attachments.UnitSupportAttachment;
+import java.util.function.Predicate;
+import lombok.AccessLevel;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.Value;
+
+/**
+ * Calculates defense strength and roll for AA/Targeted dice
+ *
+ * <p>This takes into account friendly support, and enemy support
+ */
+@Builder
+@Value
+@Getter(AccessLevel.NONE)
+class AaDefenseCombatValue implements CombatValue {
+
+  @NonNull GameData data;
+  @NonNull AvailableSupports friendlySupportTracker;
+  @NonNull AvailableSupports enemySupportTracker;
+
+  @Override
+  public StrengthOrRollCalculator getRoll() {
+    return new AaDefenseRoll(friendlySupportTracker, enemySupportTracker);
+  }
+
+  @Override
+  public StrengthOrRollCalculator getStrength() {
+    return new AaDefenseStrength(data, friendlySupportTracker, enemySupportTracker);
+  }
+
+  @Override
+  public boolean isDefending() {
+    return true;
+  }
+
+  @Override
+  public GameData getGameData() {
+    return data;
+  }
+
+  static class AaDefenseRoll extends StrengthOrRollCalculator {
```

Comment:
`StrengthOrRollCalculator` IMO is arguably a confusing concept. When is it is a strength calculator, when it is it a roll calculator. If we see an instance of it, do we know what it is or do we have to look at its state and trace back any interactions with the object to understand it. I think this hints to a SRP violation.

If we can break out "StrengthOrRollCalculator" into two interfaces, I believe that would resolve the above questions. Any thoughts/comments on options here @trevan ?

Reply:
`StrengthOrRollCalculator` is an abstract class that shares common code that both the roll and strength calculators need.  I could duplicate the class but then both classes would be identical except for their names and that's a different problem.

I could create a `StrengthCalculator` interface and a `RollCalculator` interface and give each of the child classes the corresponding interface but I don't think that will solve your issues with `StrengthOrRollCalculator` since it would still be around.

So, I'm not sure how to resolve your concerns.

Reply:
Perhaps rename `StrengthOrRollCalculator` to `StrengthAndRollCalculator`? Or let's make it a more generic term, like `CombatStrengthCalculator`